apps/builder/app/builder/shared/tree/index.tsx

```diff
@@ -59,21 +65,42 @@ export const InstanceTree = (
     [instances]
   );
 
+  const updateInstanceLabel = useCallback(
+    (instanceId: string, value: string) => {
+      store.createTransaction([instancesStore], (instances) => {
+        const instance = instances.get(instanceId);
+        if (instance === undefined) {
+          return;
+        }
+        instance.label = value;
+      });
+    },
+    []
+  );
+
   const renderItem = useCallback(
     (props: TreeItemRenderProps<Instance>) => {
       const meta = metas.get(props.itemData.component);
       if (meta === undefined) {
         return <></>;
       }
+      const label = getInstanceLabel(props.itemData, meta);
       return (
         <TreeItemBody {...props} selectionEvent="focus">
-          <TreeItemLabel prefix={<MetaIcon icon={meta.icon} />}>
-            {getInstanceLabel(props.itemData, meta)}
-          </TreeItemLabel>
+          <TreeItem
+            isEditing={props.itemData.id === editingItemId}
```

Comment:
Why you expose this state to whole builder? Can be local

Reply:
@TrySound you propose passing editingItemId instead as a prop?

Reply:
I propose to maintain "editing" state inside TreeItem

Reply:
See the issue is that both react and react-hotkeys rely on document events. So they cannot prevent each other. You can workaround this by listening keydown on element itself in effect and stopping propagation. In long term having as less as possible of global states reduce complexity of the app so it's important to not fall into "keep everything in stores".

Reply:
Yes, for `hotkeys` it's fine. But the root `event` listener is intercepting all the `keyDown` events on the tree node when the item does into edit mode. So, moving the `edit` state into local creates the issues with the root listener.

Reply:
I see basically we use that to stop listening for those other events. Maybe the solution is to add a check in those events to not react when key is pressed in an input?

Reply:
Yeah, that could be one solution. We can try that. By checking the event origin and the stop reacting to those. I will give it a try 👍

Reply:
space is not intercepted here, space triggers click event

Reply:
I might be looking in the wrong place. But, the `root` is handling the `" "` too for expanding the node here.
https://github.com/webstudio-is/webstudio-builder/blob/main/packages/design-system/src/components/tree/tree.tsx#L369

Reply:
Right, I missed that. Well, we anyway need to prevent these events while typing. Because labels may have spaces.

Reply:
Hmm shouldn't you be able to catch those events on the parent of a content editable and prevent propagation?

Though even in this case, its still kinda bad abstraction because it would have to know which events to stop propagate.

Its kinda better to let that other thing know we are editing text ... so it seems a prop or a store is unavoidable

Reply:
Yeah i kinda tried multiple variations of the same. And ended up coming to this. I tried stopping propagation too. But, some events are still bubbling as the `onKeyDown` is being triggered on the `child` node and `root` node at the same time :X

Reply:
yeah, so either prop like you did or a store that can be imported from multiple places @TrySound what's better?